src/v/cloud_storage/tests/segment_meta_cstore_test.cc

```diff
@@ -1342,3 +1344,551 @@ BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_append_retrieve_edge_case) {
+              m.committed_offset);
+        }
+        vlog(test.info, "\n\n\n\n");
+        vlog(test.info, "Expected segments:");
+        for (const auto& m : cstore) {
+            vlog(test.info, "......{}-{}", m.base_offset, m.committed_offset);
+        }
+        return false;
+    };
+
+    // Last element invariants
+    auto check_last_element_invariants = [&](const segment_meta& last_element) {
+        auto cs_last = cstore.last_segment();
+        vlog(
+          test.info,
+          "Checking last element {}, cstore last element {}",
+          last_element,
+          cs_last);
+        auto o = last_element.base_offset;
+        BOOST_REQUIRE(cond(cs_last == last_element, last_element, false));
+        // check that it can be found by its base offset
+        {
+            auto it = cstore.find(o);
+            BOOST_REQUIRE(cond(it != cstore.end(), last_element, false));
+            BOOST_REQUIRE(
+              cond(*it == cstore.last_segment(), last_element, false));
+        }
+        // check that prev(end) matches the last_segment
+        {
+            auto e = cstore.end();
+            auto it = cstore.prev(e);
+            BOOST_REQUIRE(cond(*it == last_element, last_element, false));
+        }
+    };
+
+    // Whole log invariants
+    auto check_global_invariants = [&](model::offset first) {
+        if (cstore.empty()) {
+            return;
+        }
+        auto b_it = cstore.find(first);
+        auto e_it = cstore.end();
+        BOOST_REQUIRE(cond1(b_it == cstore.begin()));
+        BOOST_REQUIRE(cond1(*cstore.prev(e_it) == cstore.last_segment()));
+    };
+
+    // Single batch invariants
+    auto check_local_invariants = [&](
+                                    const segment_meta& meta,
+                                    bool is_replacement) {
+        vlog(test.info, "Checking {}", meta);
+        auto o = meta.base_offset;
+        // check that it can be found by its base offset
+        {
+            auto it = cstore.find(o);
+            BOOST_REQUIRE(cond(it != cstore.end(), meta, is_replacement));
+            BOOST_REQUIRE(cond(*it == meta, meta, is_replacement));
+        }
+
+        // check that it can be found using upper_bound
+        {
+            auto it = cstore.upper_bound(o);
+            if (it != cstore.begin()) {
+                auto it2 = cstore.prev(it);
+                auto m2 = *it2;
+                BOOST_REQUIRE(cond(walk.is_valid(m2), m2, is_replacement));
+                BOOST_REQUIRE(cond(m2.base_offset == o, meta, is_replacement));
+            } else {
+                BOOST_REQUIRE(cond(cstore.size() == 1, meta, is_replacement));
+            }
+        }
+    };
+    for (int i = 0; i < n_iter; i++) {
+        auto [t_off, exp_so] = walk.maybe_truncate();
+        if (t_off != model::offset{}) {
+            vlog(
+              test.info, "Cstore truncate {}, expected SO: {}", t_off, exp_so);
+            cstore.prefix_truncate(t_off);
```

Comment:
Would it make sense to put this behind a probability check too? Seems like a worthwhile statespace to operate a bit on a non-truncated cstore (or a span of operations where truncations/prefix truncations do not happen)